src/stan_math_backend/Cpp.ml

```diff
@@ -0,0 +1,664 @@
+(** A set of data types representing the C++ we generate *)
+
+open Core_kernel
+
+type identifier = string [@@deriving sexp]
+
+(** C++ types *)
+type type_ =
+  | Auto
+  | Void
+  | Int
+  | Double
+  | Complex of type_
+  | TemplateType of identifier
+  | Vector of type_
+  | Array of type_ * int
+  | Type_literal of identifier  (** Used for things like Eigen::Index *)
+  | Matrix of type_ * int * int
+  | Ref of type_
+  | Const of type_
+  | Pointer of type_
+  | TypeTrait of string * type_ list
+      (** e.g. stan::promote_scalar, stan:base_type *)
+[@@deriving sexp]
+
+module Types = struct
+  (** Helpers for constructing types *)
+
+  let local_scalar = Type_literal "local_scalar_t__"
+  let std_vector t = Vector t
+  let bool = Type_literal "bool"
+  let complex s = Complex s
+  let vector s = Matrix (s, -1, 1)
+  let row_vector s = Matrix (s, 1, -1)
+  let matrix s = Matrix (s, -1, -1)
+  let string = Type_literal "std::string"
+  let size_t = Type_literal "size_t"
+  let const_ref t = Const (Ref t)
+  let str_array i = Array (Const (Pointer (Type_literal "char")), i)
```

Comment:
String is a very different type in C++ from a `const char*`

```suggestion
  let const_char_array i = Array (Const (Pointer (Type_literal "char")), i)
```

Reply:
Good point, I've done the rename you suggest.